<a href="https://colab.research.google.com/github/paolofullone/DS_Paolo/blob/main/Lambda3_Desafio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Desafio - Vaga para Cientista de Dados

**Empresa**: Lambda3 (https://www.lambda3.com.br/)

**Complexidade**: Intermediário

**Tarefa avaliada**: Machine Learning - Classificação

**Requisitos**: 
- Ferramentas Essenciais.
- Conceitos e Aplicações.
- Dominando o Python
- Python para Análise de Dados.
- Machine Learning
    - Algoritmos de Classificação
    - Técnicas de Validação
    - Tunning de Modelos

**Objetivo**: Este desafio tem como objetivo avaliar as habilidades do candidato em relação a uma tarefa simples de classificação, conforme descrito pela empresa. 
Todas as etapas do desafio podem ser resolvidos com os conhecimentos adquiridos no capítulo de Machine Learning.

**Recursos**: Para este desafio é proposto o dataset Iris (https://archive.ics.uci.edu/ml/datasets/Iris)

**Importando as bibliotecas necessárias**

In [ ]:
import pandas as pd
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import classification_report
from sklearn.model_selection import cross_val_predict

<p><span style="color:red">O dataset deve ser o Iris.</span></p>

**Importando o dataset Iris.**

In [ ]:
df = pd.read_csv("/content/drive/MyDrive/CienciaDados/ModExtra1/Lambda3/iris.csv")

**Verificando os dados e atributos da base de dados.**

In [ ]:
df.head()

sepallength  sepalwidth  petallength  petalwidth        class
0          5.1         3.5          1.4         0.2  Iris-setosa
1          4.9         3.0          1.4         0.2  Iris-setosa
2          4.7         3.2          1.3         0.2  Iris-setosa
3          4.6         3.1          1.5         0.2  Iris-setosa
4          5.0         3.6          1.4         0.2  Iris-setosa

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sepallength  150 non-null    float64
 1   sepalwidth   150 non-null    float64
 2   petallength  150 non-null    float64
 3   petalwidth   150 non-null    float64
 4   class        150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


<p><span style="color:red"> Receber as entradas específicas (escolher as variáveis pertinentes).</span></p>
<p><span style="color:red"> Justificar a escolha das variáveis.</span></p>


**Separando separando os atributos que serão utilizados como features e a classe.**

In [ ]:
X = df.drop("class",axis=1)
X.head()

sepallength  sepalwidth  petallength  petalwidth
0          5.1         3.5          1.4         0.2
1          4.9         3.0          1.4         0.2
2          4.7         3.2          1.3         0.2
3          4.6         3.1          1.5         0.2
4          5.0         3.6          1.4         0.2

In [ ]:
y = df["class"]
y.head()

0    Iris-setosa
1    Iris-setosa
2    Iris-setosa
3    Iris-setosa
4    Iris-setosa
Name: class, dtype: object

**Transformando o atributo classe para valores numéricos utilizando o LabelEncoder.**

In [ ]:
label = LabelEncoder()
label.fit(y)

LabelEncoder()

In [ ]:
y = label.transform(y)

In [ ]:
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

**Verificando as formas das variáveis.**

In [ ]:
X.shape

(150, 4)

In [ ]:
y.shape

(150,)

<p><span style="color:red"> Explicar o porque escolheu o algoritmo de classificação usado (e porque não usou outros).</span></p>

O algoritmo **K Nearest Neighbors (KNN)** foi escolhido por sua simplicidade, rápida execução e por se tratar de uma base de dados pequena com poucos dados e atributos.
<br>
Pensando na estratégia de resolver o problema proposto da melhor maneira e da forma mais simples possível, o algoritmo escolhido neste cenário apresentou bons resultados, sem a necessidade de tunning e ajustes.

**Instanciando o algoritmo KNN.**

In [ ]:
clf_knn = KNeighborsClassifier()

**Treinando o modelo utilizando a validação utilizando Cross Validation e armazenando os resultados.**

In [ ]:
resultados_knn = cross_val_predict(clf_knn, X, y, cv=5)

**Verificando os resultados através de métricas.**

In [ ]:
# 'Devolvendo' os nomes Iris-Setosa, Iris-Versicolor e Iris-Virginica a uma nova variável nomes_classes para facilitar a visualização da classification report.
nomes_classes = label.inverse_transform([0,1,2])

In [ ]:
print(classification_report(y,resultados_knn,target_names=nomes_classes))

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        50
Iris-versicolor       0.98      0.94      0.96        50
 Iris-virginica       0.94      0.98      0.96        50

    avg / total       0.97      0.97      0.97       150



<p><span style="color:red"> Apresentar a saída classificada</span></p>
<p><span style="color:red"> Justificar a performance do modelo de classificação, com base nos métodos de avaliação pertinentes.</span></p>


O Modelo de Classificação apresentou bons resultados utilizando a técnica de Validação Cruzada com um total de 5 folds.
<br><br>
Foram escolhidas as métricas **precision**, **recall** e **f1-score** para avaliar o modelo por permitir uma análise por classe, já que estamos lidando com um problema que contém 3 classes.
    
Através do valor da métrica de precisão (precision) podemos ver que o modelo sofreu mais para classificar a classe **Iris-Virginica** obtendo um percentual de **94%** que corresponde ao número de vezes que o modelo classificou corretamente os dados da respectiva classe sobre a quantidade de vezes que a classe foi predita.

Analisando a métrica de revocação (recall) o modelo teve mais dificuldade para classificar os dados da classe Iris-versicolor também com o percentual de **94%** que corresponde ao número de vezes que o modelo classificou corretamente diante do volume de dados da respectiva classe no dataset.

Por fim, a medida **F1-Score** que corresponde a média harmonica entre as duas medidas anteriormente citadas apresentou um bom resultado com o valor de **97%**.

### Trabalho adicional

- Aplicar um método de pré-processamento de dados como o MinMax Scaler e explicar o benefício da técnica.
- Utilizar o GridSearchCV para otimizar os parametros do KNN e descobrir qual a melhor configuração.
- Treinar e otimizar outros algoritmos como SVM, Arvores de Decisão, Random Forest, Naive Bayes e comparar os resultados.
- Explicar o desempenho melhor de cada algoritmo, porque ele supera outros.


### Dicas importantes 

- Tenha coragem.
- Seja verdadeiro e honesto.
- Priorize sempre a simplicidade.
- Exponha o seu ponto de vista sem medo.
- Dê o seu máximo!

# Tarefas extras propostas:


In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.model_selection import StratifiedKFold
from sklearn import model_selection
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from time import time

In [ ]:
pipelines = []
pipelines.append(('S-Scaled-LogisticRegression', Pipeline([('Scaler', StandardScaler()),('LR', LogisticRegression())]))),
pipelines.append(('MM-Scaled-LogisticRegression', Pipeline([('Scaler', MinMaxScaler()),('LR', LogisticRegression())]))),

pipelines.append(('S-Scaled-KNN', Pipeline([('Scaler', StandardScaler()),('KNN', KNeighborsClassifier())]))),
pipelines.append(('MM-Scaled-KNN', Pipeline([('Scaler', MinMaxScaler()),('KNN', KNeighborsClassifier())]))),

pipelines.append(('S-Scaled-SVM', Pipeline([('Scaler', StandardScaler()),('SVM', SVC())]))),
pipelines.append(('MM-Scaled-SVM', Pipeline([('Scaler', MinMaxScaler()),('SVM', SVC())]))),

pipelines.append(('S-Scaled-RandomForest', Pipeline([('Scaler', StandardScaler()),('RF', RandomForestClassifier())]))),
pipelines.append(('MM-Scaled-RandomForest', Pipeline([('Scaler', MinMaxScaler()),('RF', RandomForestClassifier())]))),

pipelines.append(('S-Scaled-DecisionTree', Pipeline([('Scaler', StandardScaler()),('DTC', DecisionTreeClassifier())]))),
pipelines.append(('MM-Scaled-DecisionTree', Pipeline([('Scaler', MinMaxScaler()),('DTC', DecisionTreeClassifier())]))),

pipelines.append(('S-Scaled-NaiveBayes', Pipeline([('Scaler', StandardScaler()),('NB', GaussianNB())]))),
pipelines.append(('MM-Scaled-NaiveBayes', Pipeline([('Scaler', MinMaxScaler()),('NB', GaussianNB())]))),

resultados = []
nomes = []

# Percorrendo cada um dos modelos
for nome, modelo in pipelines:
    kfold = StratifiedKFold(n_splits=10, shuffle = True, random_state=12)
    cross_val_result = model_selection.cross_val_score(modelo, 
                                                     X, 
                                                     y, 
                                                     cv = kfold, 
                                                     scoring = 'accuracy')
    resultados.append(cross_val_result)
    nomes.append(nome)
    texto = "%s: %f (%f)" % (nome, cross_val_result.mean(), cross_val_result.std())
    print(texto)

S-Scaled-LogisticRegression: 0.960000 (0.032660)
MM-Scaled-LogisticRegression: 0.933333 (0.059628)
S-Scaled-KNN: 0.953333 (0.042687)
MM-Scaled-KNN: 0.946667 (0.040000)
S-Scaled-SVM: 0.966667 (0.033333)
MM-Scaled-SVM: 0.966667 (0.033333)
S-Scaled-RandomForest: 0.960000 (0.032660)
MM-Scaled-RandomForest: 0.946667 (0.026667)
S-Scaled-DecisionTree: 0.960000 (0.032660)
MM-Scaled-DecisionTree: 0.953333 (0.030551)
S-Scaled-NaiveBayes: 0.953333 (0.042687)
MM-Scaled-NaiveBayes: 0.953333 (0.042687)


# Avaliações:
* A melhor acurácia foi do SVM, tanto para o MinMaxScaler quanto para o StandardScaler o resultado foi igual, lembrando que o dataset Iris é bem simples e balanceado.
* o MinMaxScaler transforma cada feature individualmente e as coloca em uma escala de 0 a 1.
* O StandardScaler transforma os dados para uma média próxima de zero e um desvio padrão próximo a um, ou seja, assume que os valores não são discrepantes e os transforma em uma distribuição normal.
* O benefício das técnicas é normalizar os dados das features e deixarem todos em uma mesma escala.


# Vamos escolher o SVM que teve a melhor acurácia e um desvio padrão bem próximo do menor para executar o GridSearch e o RandomizedSearch.

In [ ]:
def svc_param_selection(X, y, nfolds):
    Cs = [0.001, 0.01, 0.1, 1, 10, 100]
    gammas = [0.01, 0.1, 1, 10, 100]
    kernels = ['linear', 'rbf', 'poly']
    param_grid = {'C': Cs, 'kernel': kernels, 'gamma': gammas}
    grid_search = GridSearchCV(SVC(random_state=10), param_grid, cv=nfolds)
    grid_search.fit(X, y)
    grid_search.best_params_
    return grid_search.best_params_

In [ ]:
svc_param_selection(X, y, 10)

{'C': 0.001, 'gamma': 1, 'kernel': 'poly'}

In [ ]:
clf_svc = SVC(C=0.001, gamma=1, kernel='poly')

**Treinando o modelo utilizando a validação utilizando Cross Validation e armazenando os resultados.**

In [ ]:
resultados_svc = cross_val_predict(clf_svc, X, y, cv=5)

**Verificando os resultados através de métricas.**

In [ ]:
print(classification_report(y,resultados_svc,target_names=nomes_classes))

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        50
Iris-versicolor       0.98      0.92      0.95        50
 Iris-virginica       0.92      0.98      0.95        50

       accuracy                           0.97       150
      macro avg       0.97      0.97      0.97       150
   weighted avg       0.97      0.97      0.97       150



**Métricas antes do tunning:**

In [ ]:
print(classification_report(y,resultados_knn,target_names=nomes_classes))

                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00        50
Iris-versicolor       0.98      0.94      0.96        50
 Iris-virginica       0.94      0.98      0.96        50

       accuracy                           0.97       150
      macro avg       0.97      0.97      0.97       150
   weighted avg       0.97      0.97      0.97       150



* A única métrica que piorou após o tunning do svc comparando com o knn foi a precisão da classe Iris-Virginica.
* O recall da Iris-Versicolor melhorou de 0.92 para 0.94.
* o f1-score da Iris-Versicolor e da Iris-Virginica melhoraram de 0.95 para 0.96.

* a precisão verifica quantas vezes uma classe foi predita corretamente divido pela quantidade total de vezes que a classe foi predita.

* recall ou revocaçãp representa a quantidade de vezes que uma classe foi predita corretamente dividido pela quantidade de vezes que a classe aparece no dado de teste.

* a f1-score é  é a média harmônica entre precisão e revocação. Com essa informação podemos dizer a performance do classificador com um indicador apenas.



